pkg/ingester/ingester_v2.go

```diff
@@ -198,10 +198,17 @@ func (i *Ingester) v2Push(ctx old_ctx.Context, req *client.WriteRequest) (*clien
 		lset := client.FromLabelAdaptersToLabelsWithCopy(ts.Labels)
 
 		for _, s := range ts.Samples {
-			_, err := app.Add(lset, s.TimestampMs, s.Value)
+			// Check rate-limit
+			var err error
+			if e := i.limiter.AssertMaxSeriesPerUser(userID, int(db.currentNumSeries())); e != nil {
+				err = makeLimitError(perUserSeriesLimit, e)
+			}
```

Comment:
This assumes that every sample creates a new series. If a user has touched the limit and is trying to add a sample to an existing series, this will error out. The check should be done only when creating a new series, which seems to be tricky as the creation is internal to TSDB.

Reply:
Oh, the code seems to be wrong. When the limit is reached, the existing series cannot be pushed.

I seem to need to keep recent Series in memory or change Prometheus API...

Reply:
I think https://github.com/cortexproject/cortex/pull/2057 will help. It can tell if a series exists or no and based on that we can apply the rate limit using `NumSeries()`. The interaction with `Purge()` method would need some attention. Maybe you can wait till that PR is merged.